wstl/wstl2milp.py

```diff
@@ -0,0 +1,192 @@
+            self.model.addConstr(z <= z_child)
+        z_children, _ = zip(*vars_children)
+        self.model.addConstr(z >= 1 - len(z_children) + sum(z_children))
+
+    def disjunction(self, formula, z, rho, t):
+        '''Adds a disjunction to the model.'''
+        assert formula.op == Operation.OR
+        z_children, rho_children = zip(*[self.to_milp(f, t)
+                                         for f in formula.children])
+        z_hat_children, _ = zip(*[self.add_hat_variable(f, formula, t)
+                                 for f in formula.children])
+        vars_children = zip(z_children, z_hat_children, rho_children)
+        for k, (z_child, z_hat_child, rho_child) in enumerate(vars_children):
+            weight = formula.weight(k)
+            self.model.addConstr(
+                rho <= weight * rho_child + self.M * (1 - z_hat_child))
+            self.model.addConstr(z >= z_child)
+            self.model.addConstr(z_hat_child <= z_child)
+        self.model.addConstr(z <= sum(z_children))
+        self.model.addConstr(sum(z_hat_children) >= z)
+
+    def globally(self, formula, z, rho, t):
+        '''Adds a globally to the model.'''
+        assert formula.op == Operation.ALWAYS
+        a, b = int(formula.low), int(formula.high)
+        child = formula.child
+        vars_children = [self.to_milp(child, t + tau) for tau in range(a, b+1)]
+        for tau, (z_child, rho_child) in zip(range(a, b+1), vars_children):
+            weight = formula.weight(tau)
+            self.model.addConstr(rho <= weight * rho_child)
+            self.model.addConstr(z <= z_child)
+        z_children, _ = zip(*vars_children)
+        self.model.addConstr(z >= 1 - len(z_children) + sum(z_children))
+    
+    def eventually(self, formula, z, rho, t):
+        '''Adds an eventually to the model.'''
+        assert formula.op == Operation.EVENT
+        a, b = int(formula.low), int(formula.high)
+        child = formula.child
+        z_children, rho_children = zip(*[self.to_milp(child, t + tau)
+                                         for tau in range(a, b+1)])
+        z_hat_children, _ = zip(*[self.add_hat_variable(child, formula, t + tau)
+                                  for tau in range(a, b+1)])
+        vars_children = zip(range(a, b+1), z_children, z_hat_children,
+                            rho_children)
+        for tau, z_child, z_hat_child, rho_child in vars_children:
+            weight = formula.weight(tau)
+            self.model.addConstr(
+                rho <= weight * rho_child + self.M * (1 - z_hat_child))
+            self.model.addConstr(z_hat_child <= z_child)
+            self.model.addConstr(z >= z_child)
+        self.model.addConstr(z <= sum(z_children))
+        self.model.addConstr(sum(z_hat_children) >= z)
+
+    def until(self, formula, z, rho, t):
+        '''Adds an until to the model.'''
+        #TODO: create milp constraints for until operator
```

Comment:
Add a `TODO.md` file to the project and add an item to the list: `Until` method for the `wstl2milp` class encoding wSTL to MILP.

Reply:
Also, add an list item to the `TODO.md` for the `Release` operator.